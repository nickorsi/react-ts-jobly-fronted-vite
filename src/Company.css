.roles h3 {
  text-align: center;
}

.company-info p {
  font-size: 1.1rem;
  margin-top: 10px;
}
.card {
  background-color: white;
  border-radius: 10px;
  padding: 15px;
  margin: 15px;
}

.company-brand {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

@media screen and (max-width: 700px) {
  .company-brand {
    flex-direction: column;
  }
}

.company-brand img {
  width: 150px;
  border: solid black 2px;
  border-radius: 5px;
}

.company-container {
  width: 100%;
  max-width: 1000px;
}