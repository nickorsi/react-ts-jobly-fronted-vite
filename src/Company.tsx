import React from "react";
import { CompanyDataFromAPI } from "./CompanyList";
import { JobsDataFromAPI } from "./JobList";

type SpecificCompanyDataFromAPI = CompanyDataFromAPI & {jobs: JobsDataFromAPI}

function Company() {
    return (
        <div>Company Component</div>
    )
}

export {Company};
export type {SpecificCompanyDataFromAPI};