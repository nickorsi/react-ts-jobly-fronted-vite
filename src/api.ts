const BASE_URL:string = import.meta.env.VITE_REACT_APP_BASE_URL || "http://localhost:3001";

type CompanyDataToAPI = {
  handle?: string,
  name: string,
  description: string,
  numEmployees: number,
  logoURL: string,
};

type JobDataToAPI = {
  title: string,
  salary: number,
  equity: number,
  companyHandle?: string,
};

type UserDataToAPI = {
  username?: string,
  password?: string,
  firstName?: string,
  lastName?: string,
  email?: string,
  isAdmin?: boolean,
}

type FilterDataToAPI = {
  nameLike?: string,
  title?: string,
}

type Data = CompanyDataToAPI | UserDataToAPI | JobDataToAPI | UserDataToAPI | FilterDataToAPI | Record<string, never>;

/** API Class.
 *
 * Static class tying together methods used to get/send to the API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't
 * be any API-aware stuff elsewhere in the frontend.
 *
 */

class JoblyApi {
  // Remember, the backend needs to be authorized with a token
  // We're providing a token you can use to interact with the backend API
  // DON'T MODIFY THIS TOKEN
  static token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZ" +
    "SI6InRlc3R1c2VyIiwiaXNBZG1pbiI6ZmFsc2UsImlhdCI6MTU5ODE1OTI1OX0." +
    "FtrMwBQwe6Ue-glIFgz_Nf8XxRT2YecFCiSpYL0fCXc";

  static async request(endpoint:string, data:Data = {}, method:string = "GET") {
    const url = new URL(`${BASE_URL}/${endpoint}`);
    const headers = {
      authorization: `Bearer ${JoblyApi.token}`,
      'content-type': 'application/json',
    };

    url.search = (method === "GET")
      ? new URLSearchParams(data as FilterDataToAPI).toString()
      : "";

    // set to undefined since the body property cannot exist on a GET method
    const body = (method !== "GET")
      ? JSON.stringify(data)
      : undefined;

    const resp = await fetch(url, { method, body, headers });

    if (!resp.ok) {
      console.error("API Error:", resp.statusText, resp.status);
      const message = (await resp.json()).error.message;
      throw Array.isArray(message) ? message : [message];
    }

    return await resp.json();
  }

  // Individual API routes

  /** Get details on a company by handle. */

  static async getCompany(handle: string) {
    const res = await this.request(`companies/${handle}`);
    return res.company;
  }


  /**
   * Get details on companies based on passed in searchTerm string.
   * Note: If searchTerm is empty, returns all companies.
  */

  /**
   * Get details on jobs based on passed in searchTerm string.
   * Note: If searchTerm is empty, returns all jobs.
  */

  static async getJobs(searchTerm: string = '') {
    const data:FilterDataToAPI = {};
    if(searchTerm.length > 0) {
       data['title'] = searchTerm;
    }
    const res = await this.request(`jobs`, data);
    return res.jobs;
  }
}

export { JoblyApi, BASE_URL };

export type {JobDataToAPI, CompanyDataToAPI};
