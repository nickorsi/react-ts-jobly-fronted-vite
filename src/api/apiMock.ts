import { http, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';

type methodOptions = 'get' | 'post' | 'patch' | 'delete';

type dataType = {
  [key:string]: any
}

const mockServer = setupServer();

function mockAPI(
  method: methodOptions,
  path:string,
  data: dataType,
  status:number = 200
) {

mockServer.use(http[method](path, () => {
    // At one point received a remove stderror from msw for redundent
    // query-parameters being present in the passed in path parameter.
    // These queries aren't really relied on in the return value because
    // the return data is already mocked according to the passed in query.
    // At a later time it appears this stderror is no longer occuring, but
    // left this code in case it pop up again.
    // const url = new URL(request.url); // request object needs to be destructured in the param of the arrow function.

    // const nameLike = url.searchParams.get('nameLike');
    // const title = url.searchParams.get('title');

    return HttpResponse.json(data, {status: status})
    })
  )
}

export {mockAPI, mockServer};