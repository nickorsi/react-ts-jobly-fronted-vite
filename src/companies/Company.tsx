import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import { CompanyDataFromAPI } from "./CompanyList";
import { JobsDataFromAPI } from "../jobs/JobList";
import { JobCardList } from "../jobs/JobCardList";
import LoadingSpinner from "../common/LoadingSpinner";
import ErrorMessage from "../common/ErrorMessage";
import { JoblyApi } from "../api/api";
import "./Company.css";

type SpecificCompanyDataFromAPI = CompanyDataFromAPI & {jobs: JobsDataFromAPI}

/**
 * Company componenet renders a page with company data and descendent
 * componenets related to jobs associated with the company. Holds data related
 * to the company in state. Retrieves data on mount.
 *
 * Rendered by RoutesList on the /companies/:companyHandle path
 *
 * RoutesList -> Company -> JobList
 */
function Company() {
    const { companyHandle }  = useParams();
    const [companyData, setCompanyData] = useState<SpecificCompanyDataFromAPI | null>(null);
    const [errors, setErrors] = useState<string[]>([]);

    useEffect(function () {
        async function getCompanyData(companyHandle: string) {
            try{
                const companyData = await JoblyApi.getCompany(companyHandle);
                setCompanyData(companyData);
                setErrors([]);
            } catch (error) {
                setErrors(error as string[]);
            }
        }
        getCompanyData(companyHandle as string);
    }, []);

    if(errors.length > 0) {
        return (
            <ErrorMessage errors={errors}></ErrorMessage>
        )
    }

    if(companyData === null) {
        return (
            <LoadingSpinner></LoadingSpinner>
        )
    }


    return (
        <div className="company-container">
            <div className="company-info card">
                <div className="company-brand">
                    <h1>{ companyData.name }</h1>
                    {
                        companyData.logoUrl === null ? //comparing to null object
                        null :
                        <img src={`${companyData.logoUrl}`} alt="company logo" />
                    }
                </div>
                <p>{companyData.description}</p>
            </div>
            <h2>Open Positions</h2>
            {companyData.jobs.length > 0 ?
                <JobCardList jobs={companyData.jobs}></JobCardList> :
                <p className="no-results">Sorry, there are no open roles at this time.</p>
            }
        </div>
    )
}

export {Company};
export type {SpecificCompanyDataFromAPI};