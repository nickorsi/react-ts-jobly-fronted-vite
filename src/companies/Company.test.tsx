import {describe, expect, it, vi} from "vitest";
import { render, waitFor } from "@testing-library/react";

import { JoblyApi } from "../api/api.ts";
import { Company } from "./Company.tsx";

/**
 * Something wrong with test config, uncomment line below to get intellisense to
 * recognize the extended expect methods, comment out for tests to run properly.
*/
// import "@testing-library/jest-dom";

const mockedGetCompany = vi.spyOn(JoblyApi, "getCompany");

describe("Simple Render Tests", () => {
    it("renders without crashing", function () {
        render(<Company />);
    });
    it("matches the snapshot", function () {
        const { container } = render(
            <Company />
        );

        expect(container).toMatchSnapshot();
    });
});

describe("Different Rendered Condition Tests", () => {
    it("renders loading component", function () {
        const { container } = render(<Company />);

        expect(mockedGetCompany).toHaveBeenCalledOnce();
        expect(container).toContainHTML("Loading...");
    });

    it("renders company info", async function () {
        mockedGetCompany.mockResolvedValue({
            handle: "c1",
            name: "Company 1",
            description: "Description 1",
            numEmployees: 1,
            logoUrl: "http//c1.img",
            jobs: [
              {
                id: 1,
                title: "Job 1",
                salary: 1,
                equity: "1",
                companyName: "Company 1",
              }
            ]
        });

        const { container } = render(<Company />);

        expect(mockedGetCompany).toHaveBeenCalledOnce();

        await waitFor(() => {
            expect(container).toContainHTML("Company 1");
        });
    });

    it("renders error if company not found", async function () {
        mockedGetCompany.mockRejectedValue(
            ['Test Error']
        );

        const { container } = render(<Company />);

        expect(mockedGetCompany).toHaveBeenCalledOnce();

        await waitFor(() => {
            expect(
                container.querySelector(".error-container p")?.textContent
            ).toEqual(
                "Looks like something went wrong: Test Error"
            );
        });
    })
});