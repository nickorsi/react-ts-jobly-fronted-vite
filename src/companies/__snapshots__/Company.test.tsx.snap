// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Simple Render Tests > matches the snapshot 1`] = `
<div>
  <div
    class="spinner d-flex align-items-center"
  >
    <div
      aria-hidden="true"
      class="spinner-border me-3 text-light"
    />
    <strong
      class="text-light"
      role="status"
    >
      Loading...
    </strong>
  </div>
</div>
`;
